### PR TITLE
[WFCORE-433]: git backend for loading/storing the configuration XML for wildfly

### PR DESCRIPTION
**Pull request by ehsavoie**
Having it working properly in standalone.
Deleting content of the configuration directory before cloning so we have the logging.properties on boot.
Fixing too many commits on boot
Allowing for git repository creation with the special repo 'local'.
Snapshots can take a comment that will be used as commit comment.
Adding support for managed content.
Removing domain support
Using tags for snapshot management
Integrating Elytron

Jira: https://issues.jboss.org/browse/WFCORE-433
JBEAP: https://issues.jboss.org/browse/EAP7-728
Proposal PR: https://github.com/wildfly/wildfly-proposals/pull/80
Documentation PR: https://github.com/wildfly/wildfly/pull/11492

### CODE REVIEW

**Comment by jmesnil** [comment]
@ehsavoie Code is not compiliing:

```
[Step 2/5] [ERROR] /store/work/tc-work/37b47ae8b9c60325/core/server/src/main/java/org/jboss/as/server/logging/ServerLogger.java:[1317,10] Message id 269 is not unique for messageMethod failedToPublishConfiguration with project code WFLYSRV.
[07:30:01][Step 2/5] [ERROR] /store/work/tc-work/37b47ae8b9c60325/core/server/src/main/java/org/jboss/as/server/logging/ServerLogger.java:[1333,10] Message id 269 is not unique for messageMethod errorUsingGit with project code WFLYSRV.
```

File: server/src/main/java/org/jboss/as/server/deployment/DeploymentAddHandler.java [hunk 1]
```diff
@@ -33,6 +33,7 @@
 import static org.jboss.as.server.controller.resources.DeploymentAttributes.RUNTIME_NAME;
 import static org.jboss.as.server.controller.resources.DeploymentAttributes.SERVER_ADD_ATTRIBUTES;
 import static org.jboss.as.server.deployment.DeploymentHandlerUtils.asString;
+import static org.jboss.as.server.deployment.DeploymentHandlerUtils.createFailureException;
```
[the remark this answers is not in the document]
Comment (ehsavoie) [reply server/src/main/java/org/jboss/as/server/deployment/DeploymentAddHandler.java]: How so ?

File: server/src/main/java/org/jboss/as/server/controller/git/GitContentRepository.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+                    git.commit().setMessage(message).setAmend(true).setAll(true).setNoVerify(true).call();
+                }
+            } catch (RevisionSyntaxException | IOException | GitAPIException ex) {
+                throw new RuntimeException(ex);
+            }
+        }
+    }
+
+    public static void addService(final ServiceTarget serviceTarget, final GitRepository gitRepository, final File repoRoot, final File tmpRoot) {
+        if (gitRepository != null) {
```
[the remark this answers is not in the document]
Comment (ehsavoie) [reply server/src/main/java/org/jboss/as/server/controller/git/GitContentRepository.java]: nope

File: controller/src/main/java/org/jboss/as/controller/persistence/ConfigurationFile.java [hunk 3]
```diff
@@ -590,8 +597,9 @@ private void moveFile(final File file, final File backup) throws IOException {
         Files.move(file.toPath(), backup.toPath(), StandardCopyOption.REPLACE_EXISTING);
     }
 
-    String snapshot() throws ConfigurationPersistenceException {
-        String name = getTimeStamp(new Date()) + mainFile.getName();
+    String snapshot(String prefix, String comment) throws ConfigurationPersistenceException {
+        String fileName = (comment == null || comment.isEmpty()) ? mainFile.getName() : comment + "-" + mainFile.getName();
```
[the remark this answers is not in the document]
Comment (ehsavoie) [reply controller/src/main/java/org/jboss/as/controller/persistence/ConfigurationFile.java]: Not sure about that , what would be the reason ?

File: controller/src/main/java/org/jboss/as/controller/persistence/AbstractFilePersistenceResource.java [hunk 4]
```diff
@@ -55,13 +57,21 @@ public void commit() {
         if (marshalled == null) {
             throw ControllerLogger.ROOT_LOGGER.rollbackAlreadyInvoked();
         }
-        doCommit(marshalled);
+        try(InputStream in = getMarshalledInputStream()) {
+            doCommit(in);
+        } catch (IOException ioex) {
+           ioex.printStackTrace();
```
[the remark this answers is not in the document]
Comment (ehsavoie) [reply controller/src/main/java/org/jboss/as/controller/persistence/AbstractFilePersistenceResource.java]: agreed

File: controller/src/main/java/org/jboss/as/controller/operations/common/SnapshotTakeHandler.java [hunk 5]
```diff
@@ -64,9 +72,12 @@ public void execute(OperationContext context, ModelNode operation) throws Operat
         if (authorizationResult.getDecision() == AuthorizationResult.Decision.DENY) {
             throw ControllerLogger.ROOT_LOGGER.unauthorized(operation.get(OP).asString(), context.getCurrentAddress(), authorizationResult.getExplanation());
         }
-
+        ModelNode commentNode = COMMENT.resolveModelAttribute(context, operation);
+        ModelNode snapshotNode = SNAPSHOT_NAME.resolveModelAttribute(context, operation);
+        String comment = commentNode.isDefined() ? commentNode.asString() : null;
+        String snapshot = snapshotNode.isDefined() ? snapshotNode.asString() : null;
```
[the remark this answers is not in the document]
Comment (ehsavoie) [reply controller/src/main/java/org/jboss/as/controller/operations/common/SnapshotTakeHandler.java]: ok

File: controller/pom.xml [hunk 6]
```diff
@@ -92,6 +92,11 @@
             <artifactId>vdx-wildfly</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>org.eclipse.jgit</groupId>
```
[the remark this answers is not in the document]
Comment (ehsavoie) [reply controller/pom.xml]: ok

File: controller/src/main/java/org/jboss/as/controller/persistence/ConfigurationPersister.java [hunk 7]
```diff
@@ -107,8 +111,33 @@ default boolean isPersisting() {
...
+    }
+
+    /**
+     * Take a snapshot of the current configuration.
+     *
+     * @param message optionnal message
+     * @return the file location of the snapshot
+     * @throws ConfigurationPersistenceException if a problem happened when creating the snapshot
+     */
+    default String snapshot(String name, String message) throws ConfigurationPersistenceException {
```
[the remark this answers is not in the document]
Comment (ehsavoie) [reply controller/src/main/java/org/jboss/as/controller/persistence/ConfigurationPersister.java]: fixed

File: server/src/main/java/org/jboss/as/server/deployment/DeploymentAddHandler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jmesnil) [reply server/src/main/java/org/jboss/as/server/deployment/DeploymentAddHandler.java]: the only change in this file is reorganizing the imports. It's adding noise to the patch.

File: controller/src/main/java/org/jboss/as/controller/persistence/ConfigurationFile.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmesnil) [reply controller/src/main/java/org/jboss/as/controller/persistence/ConfigurationFile.java]: what would happen if `comment` contains new lines and we end up with them in the file name?

File: controller/src/main/resources/org/jboss/as/controller/descriptions/common/LocalDescriptions.properties [hunk 8]
```diff
@@ -623,6 +623,9 @@ read-config-as-xml=Reads the current configuration and returns it in XML format.
 read-config-as-xml.response=The XML form of the persistent configuration.
 add-deployer-chains=Adds deployer chain
 
+#Publish
+publish.publish-configuration=Publish the configuration to some external location.
+publish.publish-configuration.location=The location where the configuration is going to be published.
```
[the remark this answers is not in the document]
Comment (ehsavoie) [reply controller/src/main/resources/org/jboss/as/controller/descriptions/common/LocalDescriptions.properties]: it is a valid remote git entry. 'local' is not supported

**Comment by ehsavoie** [comment]
retest this pealse

**Comment by ehsavoie** [comment]
retest this please

**Comment by ehsavoie** [comment]
retest this please

**Comment by ehsavoie** [comment]
retest this please

**Comment by kabir** [comment]
@ehsavoie needs a rebase

**Comment by ehsavoie** [comment]
retest this please

**Comment by ehsavoie** [comment]
retest this please

**Comment by ehsavoie** [comment]
retest this please

**Comment by ehsavoie** [comment]
retest this please

**Comment by ehsavoie** [comment]
Retest this please

**Comment by ehsavoie** [comment]
retest this please

**Comment by jmesnil** [comment]
@wildfly/mergers-core I removed the missing-reqs label as all requirements are met.

@ehsavoie, this PR will be ready to merge once the failures on the CI has been fixed.

**Pull request merged**